CCP-176 Pagination using lucene for conceptsearch

mkarthik90 opened the pull request.

diging-jenkins commented:
Can one of the admins verify this patch?

mkarthik90 commented:
@jdamerow  Added the changes for CCP-175

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/core/impl/IndexService.java:
```diff
@@ -53,10 +53,26 @@
-
-		return luceneUtility.queryIndex(fieldMap, operator);
+        // Fetches all the pages
+        return luceneUtility.queryIndex(fieldMap, operator, Integer.MIN_VALUE, Integer.MIN_VALUE);
```
why don't you put in 0 for start index and -1 for page number? and then have logic in your lucene utility that checks for -1? This seems odd and not very intuitive.

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/rest/ConceptSearch.java:
```diff
@@ -55,29 +60,53 @@
-    public @ResponseBody ResponseEntity<String> searchConcept(HttpServletRequest req) {
-        Map<String, String[]> queryParams = req.getParameterMap();
+    public @ResponseBody ResponseEntity<String> searchConcept(
+            @Valid ConceptSearchParameters conceptSearchParameters)
```
It's a good idea to validate those parameters, but there doesn't seem to be a validator?

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/rest/ConceptSearch.java:
```diff
@@ -55,29 +60,53 @@
+            searchResults = manager.searchForConceptByPageNumberAndFieldMap(searchFields, operator,
+                    page, numberOfRecordsPerPage);
         } catch (LuceneException | IllegalAccessException | IndexerRunningException ex) {
             return new ResponseEntity<String>(ex.getMessage(), HttpStatus.BAD_REQUEST);
```
Please respond with correct error codes here.

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/rest/ConceptSearch.java:
```diff
@@ -55,29 +60,53 @@
+        if (searchResults.length == 0) {
+            // Returning a 404 error.
+            return new ResponseEntity<String>("No records found for the search condition.",
+                    HttpStatus.NOT_FOUND);
```
shouldn't be not found. Not found would mean the resource wasn't found, but it was. It just didn't yield any results, which is just as ok as results.

mkarthik90 commented:
@jdamerow Addressed review comments

diging-jenkins commented:
Can one of the admins verify this patch?

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/rest/ConceptSearch.java:
```diff
@@ -97,14 +99,16 @@
         if (searchResults.length == 0) {
-            // Returning a 404 error.
+            // Data is not found but still returning OK
             return new ResponseEntity<String>("No records found for the search condition.",
```
Let's return some proper xml here as well. Otherwise, it would throw clients off when they expect xml but just get text.

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/rest/ConceptSearch.java:
```diff
@@ -126,4 +130,10 @@
     }
+
+    @ExceptionHandler({ BindException.class })
+    public ResponseEntity<String> handleMethodArgumentTypeMismatch(BindException ex) {
```
where is that being thrown?

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/rest/ConceptSearchParameters.java:
```diff
@@ -0,0 +1,129 @@
+    private String type_id;
+    private String creator;
+    private String modified_by;
+    private Integer number_of_records_per_page;
```
you probably want to validate a little bit more than just the word. what for example about pos value or operator?

diging-jenkins commented:
Can one of the admins verify this patch?

mkarthik90 commented:
@jdamerow Addressed review comments

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/rest/ConceptSearchParameterValidator.java:
```diff
@@ -0,0 +1,42 @@
+import edu.asu.conceptpower.servlet.core.POS;
+
+@Component
+public class ConceptSearchParameterValidator implements Validator {
```
javadoc

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/rest/ConceptSearchParameterValidator.java:
```diff
@@ -0,0 +1,42 @@
+    public void validate(Object target, Errors errors) {
+        ConceptSearchParameters conceptSearchParameter = (ConceptSearchParameters) target;
+
+        if (conceptSearchParameter.getWord() == null || conceptSearchParameter.getWord().isEmpty()) {
```
you can use validationutils here (it becomes a one liner then)

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/rest/ConceptSearchParameterValidator.java:
```diff
@@ -0,0 +1,42 @@
+                    || SearchParamters.OP_AND.equalsIgnoreCase(conceptSearchParameter.getOperator())) {
+                errors.reject("operator", "Please enter correct operator value.");
+            }
+        }
```
I think we are not using localized messages for CP yet, but lets start doing it. I'll create a new ticket for that though. So for now, let's leave it the way it is.

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/xml/XMLConstants.java:
```diff
@@ -37,4 +38,5 @@
 	public final static String MATCHES = "matches";
     public final static String NUMBER_OF_RECORDS = "number_of_records";
+    public final static String ERROR_MESSAGE = "error_message";
 }
```
let's reformat this before merging

mkarthik90 commented:
@jdamerow  Addressed review comments

diging-jenkins commented:
Can one of the admins verify this patch?

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/rest/ConceptSearchParameterValidator.java:
```diff
@@ -2,10 +2,19 @@
+/**
+ * This class validates the Concept power search parameters. Checks whether user
+ * has submitted word and pos details for searching the concept In addition, the
+ * validator checks for the operators (OR and AND)
```
- Conceptpower is one word. 
- Should be "Checks whether **a** user has submitted word and pos details **when** searching for a concept**.**" (period is missing at the end)
- periods is missing at the end of last sentence as well.

jdamerow commented:
Almost there.

mkarthik90 commented:
@jdamerow  Addressed review comments

diging-jenkins commented:
Can one of the admins verify this patch?

jdamerow commented:
ok to test

jdamerow commented:
hm, well something is wrong:
https://diging-dev.asu.edu/conceptpower-review/rest/ConceptSearch?word=pony&pos=noun&operator=and
should give me a result.

Also, if you leave out operator it throws an error, which it shouldn't because it should use the default operator.

mkarthik90 commented:
@jdamerow  Fixed review comments.

Operator by default is or
If operator is other than "and" or "or", the system throws error.

diging-jenkins commented:
Can one of the admins verify this patch?

mkarthik90 commented:
@jdamerow Merged with develop and addressed review comments

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/rest/ConceptSearch.java:
```diff
@@ -123,10 +127,15 @@ private void initBinder(WebDataBinder binder) {
-            return new ResponseEntity<String>(e.getMessage(), HttpStatus.SERVICE_UNAVAILABLE);
+        } catch (IllegalAccessException iae) {
+            logger.error("Illegal access exception", iae);
+            return new ResponseEntity<String>(iae.getMessage(), HttpStatus.BAD_REQUEST);
```
when is this exception thrown exactly?

mkarthik90 replied:
That has not been thrown. I added it without noticing that. Removed that catch statements.

jdamerow commented:
Where does it say what page is returned? This only seems to say how many results in total:
```<digitalHPS:pagination number_of_records="100" ></digitalHPS:pagination>```

mkarthik90 commented:
@jdamerow Addressed review comments

diging-jenkins commented:
Can one of the admins verify this patch?

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/rest/ConceptSearch.java:
```diff
@@ -59,37 +75,73 @@
+    public @ResponseBody ResponseEntity<String> searchConcept(
+            @Validated ConceptSearchParameters conceptSearchParameters,
+            BindingResult result)
+            throws IllegalArgumentException, IllegalAccessException {
```
Why are those thrown here? Are they caught anywhere?

mkarthik90 replied:
In this method we are using reflection technique to fetch the searchFields value. So it throws these exceptions.

jdamerow replied:
Then you need to handle that in your controller and return a proper error message. I don't think an exception thrown here would return proper xml or json and error code, would it?

mkarthik90 replied:
yes I will do that

mkarthik90 commented:
@jdamerow  Addressed review comments. 

Added global exception handlers

diging-jenkins commented:
Can one of the admins verify this patch?

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/util/ExceptionHandler.java:
```diff
@@ -36,5 +36,17 @@ public String handle(Exception e) {
 	}
+	
+    @org.springframework.web.bind.annotation.ExceptionHandler(value = IllegalArgumentException.class)
+    public ResponseEntity<String> handleIllegalArgumentException(IllegalArgumentException ex) {
```
Please find me when you're back from lunch to about it.

mkarthik90 commented:
@jdamerow  Addressed review comments.

This change has the following changes. 
1) Changes for package structure for implementing different
exception handlers. 
2) Added the code changes of CCP-171 into CCP-176. 
3) Review comment changes of CCP-171.

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/util/ExceptionHandler.java:
```diff
@@ -0,0 +1,28 @@
+	
+    private static final Logger logger = LoggerFactory.getLogger(ExceptionHandler.class);
+
+    @org.springframework.web.bind.annotation.ExceptionHandler(value = { Exception.class, IllegalArgumentException.class,
```
I"m pretty sure this has to go into the web package?

jdamerow commented:
There is a problem with the messages:
https://diging-dev.asu.edu/conceptpower-review/home/conceptsearch?word=John+Christopher+Willis+&pos=noun

jdamerow commented:
The search on the edit concept page doesn't return anything.
![image](https://cloud.githubusercontent.com/assets/8881141/20156710/b7403754-a68e-11e6-9f95-199173b37c7f.png)

jdamerow commented:
And the search for creating wordnet wrappers doesn't work anymore either.

jdamerow commented:
On the edit concept page: hitting the enter key in the search for wordnet concepts doesn't submit the search (it just does nothing).

mkarthik90 commented:
@jdamerow  Can you deploy this change in develop. I have added some logger debug statements to find the issue. I doubt if the numberOfLuceneResults from property file is not getting loaded.

diging-jenkins commented:
Can one of the admins verify this patch?

jdamerow commented:
@mkarthik90 should be deployed now

mkarthik90 commented:
@jdamerow  Addressed review comments.

Changes made only for concept list. If we make changes to all the string fields while indexing, it causes problems in other fields. For example if we index a wordnet id field as string field and convert it to lower case, we face problem while we try to fetch the data from wordnet manager, because wordnet manager checks whether wordnet id is of the form "WID". In our case it will be "wid". So to avoid confusion I just made change to concept list.

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/ConceptManager.java:
```diff
@@ -102,15 +101,31 @@ public ConceptEntry getWordnetConceptEntry(String wordnetId) throws LuceneExcept
      */
     @Override
-    public ConceptEntry[] getConceptListEntriesForWord(String word, String pos, String conceptList)
+    public ConceptEntry[] getConceptListEntriesForWordPOSConceptList(String word, String pos, String conceptList)
```
given that the method is already called get**ConceptListEntries**ForWord... it seems doppeltgemoppelt (doublemoppeled) to attach ConceptList at the end of the method.

mkarthik90 replied:
@jdamerow Can I change it as getConceptListEntriesByWordAndPosAndConceptList?

jdamerow commented:
The description needs to be shown for wordnet concepts in the edit concept screen. Otherwise it's hard to find out if the concept you're looking at is actually the one you need.

mkarthik90 commented:
@jdamerow  Addressed review comments

diging-jenkins commented:
Can one of the admins verify this patch?

jdamerow commented:
The messages are still broken:
https://diging-dev.asu.edu/conceptpower-review/home/conceptsearch?word=John+Christopher+Willis+&pos=noun

jdamerow commented:
The form should use the whole width of the page:
![image](https://cloud.githubusercontent.com/assets/8881141/20312306/38b68998-ab10-11e6-9655-c70e350d6275.png)

mkarthik90 commented:
@jdamerow Addressed review comments.

1) Fixed alignment issue
2) Fixed error message issue

diging-jenkins commented:
Can one of the admins verify this patch?

jdamerow commented:
The labels don't need that much space:
![image](https://cloud.githubusercontent.com/assets/8881141/20318286/e0f4c19a-ab26-11e6-9949-16009484f5c8.png)

jdamerow commented on Conceptpower+Spring/src/main/java/edu/asu/conceptpower/core/ConceptEntry.java:
```diff
@@ -10,12 +10,12 @@
+import edu.asu.conceptpower.app.reflect.LuceneField;
+import edu.asu.conceptpower.app.reflect.SearchField;
+import edu.asu.conceptpower.rest.LuceneFieldNames;
+import edu.asu.conceptpower.rest.SearchFieldNames;
```
Those should not be in the rest package.

mkarthik90 commented:
@jdamerow  Addressed review comments.

1) Changes for request mapping of rest interface.
2) Adjusted label and text fields.

diging-jenkins commented:
Can one of the admins verify this patch?

jdamerow commented:
There are some issues here with the search. I'll merge it for now, because of the package structure change, but we need to fix those issues before we can deploy to production. Also, please test the whole application for more issues.
